janusgraph-backend-testutils/src/main/java/org/janusgraph/graphdb/IndexMetricTest.java

```diff
@@ -75,8 +75,10 @@ public void testIndexMetrics() {
         Assertions.assertThrows(JanusGraphException.class,  () ->
             graph.indexQuery("idx", "!@#$%^").vertexTotals());
 
-        verifyIndexMetrics("search", METRICS_INDEX_PREFIX, ImmutableMap.of(M_MUTATE, 1L, M_QUERY, 1L, M_MIXED_AGG_QUERY, 5L, M_TOTALS, 2L, M_RAW_QUERY, 1L));
-        assertEquals(1, metric.getCounter(METRICS_INDEX_PREFIX, "search", M_TOTALS, MetricInstrumentedStore.M_EXCEPTIONS).getCount());
+        verifyIndexMetrics("search", METRICS_INDEX_PREFIX, ImmutableMap.of(M_MUTATE, 1L, M_QUERY, 1L,
+            M_MIXED_AGG_QUERY, 1L, M_TOTALS, 2L, M_RAW_QUERY, 1L));
```

Comment:
To be honest, it's not clear to me at all why `M_MIXED_AGG_QUERY` is now 1 instead of 5. Does this really mean that JanusGraph only performed one query to the index backend for the 5 max, min, sum, mean, and max traversals?
@rngcntr: Maybe you have a better understanding of this, given that you implemented the limited batch processing?

Reply:
I can't give an immediate answer to this either. After all, batching should not work across the boundaries of traversals :stuck_out_tongue: 
This could be utter nonsense, but I vaguely remember the elasticsearch API returning different kinds of aggregations (e.g. max, min, avg) even though a query only requires one of them. Together with result caching, that could in theory explain the first four queries being answered by a single index query. **However**, that does neither explain the fifth query, nor does should this behavior have changed with the introduction of batch processing.

Reply:
`p4` however isn't backed by an index. So, it shouldn't use a `MIXED_AGG_QUERY` for this traversal anyway, right?

Overall, I'm not sure to be honest what this test should exactly test. Do we expect to get certain metrics back or do we just want to test that we get somehow meaningful values for the different metrics?
To me, it looks more like the latter and I would therefore be OK with changing the test like this.
If we however want to verify that we get certain metrics back, then we might need to disable batch processing for just this test (?)

Reply:
- Neither LuceneIndex (which is used for this test) nor ElasticSearch (not used for this test) returns multiple aggregations. They return a single aggregation you used.
- For this test the MixedAggQuery aggregation optimization isn't used for min/max/mean/sum when query batch is enabled, but they are used when query batch is disabled.

I didn't research the reason why the optimization is not used for `min/max/mean/sum` when query.batch is enabled, but below are the profiles (for `count` query optimization is used, but for `min/max/mean/sum` it's not used).

Profile for `graph.traversal().V().has("p1", "value").values("p3").max().profile().next()` (query.batch disabled):
```
Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphMixedIndexAggStep                                            1           1          24.265   100.00
  MixedIndexCountQuery                                                                        21.351
    \_query=[(p1 = value)]:idx
                                            >TOTAL                     -           -          24.265        -
```

Profile for `graph.traversal().V().has("p1", "value").values("p3").max().profile().next()` (query.batch enabled):
```
Traversal Metrics
Step                                                               Count  Traversers       Time (ms)    % Dur
=============================================================================================================
JanusGraphStep([],[p1.eq(value)])                                      1           1           4.598    41.20
  constructGraphCentricQuery                                                                   0.578
  GraphCentricQuery                                                                            3.974
    \_condition=(p1 = value)
    \_orders=[]
    \_isFitted=true
    \_isOrdered=true
    \_query=[(p1 = value)]:idx
    \_index=idx
    \_index_impl=search
JanusGraphMultiQueryStep                                               1           1           0.832     7.46
NoOpBarrierStep(null)                                                  1           1           0.195     1.75
JanusGraphPropertiesStep([p3],value)                                   1           1           3.476    31.16
  \_condition=type[p3]
  \_orders=[]
  \_isFitted=true
  \_isOrdered=true
  \_query=p3:SliceQuery[0x50E0,0x50E1)
  \_multi=true
  \_vertices=1
  optimization                                                                                 0.317
MaxGlobalStep                                                          1           1           2.056    18.43
                                            >TOTAL                     -           -          11.159        -
```

Reply:
Thanks for investigating and the detailed explanation, @porunov! Your explanation made it quite clear why the batch processing (implemented in the `JanusGraphLocalQueryOptimizerStrategy`) effectively disabled the mixed index aggregations: The `JanusGraphMultiQueryStep` that was added by the `JanusGraphLocalQueryOptimizerStrategy` also includes the aggregation query that we can perform with the mixed index backend, but the mixed index aggregation optimization isn't able to work on a `JanusGraphMultiQueryStep`.
I fixed this by making sure that the mixed index aggregation optimizations are always applied before the `JanusGraphLocalQueryOptimizerStrategy`.